src/java.base/share/classes/java/nio/X-Buffer.java.template

```diff
@@ -1447,6 +1452,97 @@ public abstract sealed class $Type$Buffer
+     * given {@code start} index and at the given {@code index} in this buffer.
+     * The position of this buffer is unchanged.
+     *
+     * <p> In other words, an invocation of this method of the form
+     * <code>dst.put(index,&nbsp;csq,&nbsp;start,&nbsp;end)</code>
+     * has exactly the same effect as the loop
+     *
+     * <pre>{@code
+     *     for (int i = start, j = index; i < end; i++, j++)
+     *         dst.put(j, csq.charAt(i));
+     * }</pre>
+     *
+     * except that it first checks that there is sufficient space in this
+     * buffer and it is potentially much more efficient.
+     *
+     * @apiNote
+     * Unlike the other four-parameter absolute bulk put methods in this class,
+     * this method:
+     * <ul>
+     * <li>defines the range of characters to be read from the source in terms
+     * of the half-open interval {@code [start,}&nbsp;{@code  end)} instead of
+     * an {@code offset} and a {@code length};</li>
+     * <li>throws a {@code BufferOverflowException} instead of an
+     * {@code IndexOutOfBoundsException} for the condition
+     * {@code index}&nbsp;{@code +}&nbsp;{@code end}&nbsp;{@code -}&nbsp;
+     * {@code start}&nbsp;{@code >}&nbsp;{@code limit()}.</li>
+     * </ul>
+     *
+     * @param  index
+     *         The index in this buffer at which the first char will be written;
+     *         must be non-negative and less than {@code limit()}
+     *
+     * @param  csq
+     *         The {@code CharSequence} from which $type$s are to be read
+     *
+     * @param  start
+     *         The offset within the {@code CharSequence} of the first $type$
+     *         to be read; must be non-negative and no larger than
+     *         {@code csq.length()}
+     *
+     * @param  end
+     *         The offset within the {@code CharSequence} of the last $type$
+     *         to be read, plus one; must be non-negative and no larger than
+     *         {@code csq.length()}
+     *
+     * @return  This buffer
+     *
+     * @throws  BufferOverflowException
+     *          If there is insufficient space in this buffer
+     *
+     * @throws  IndexOutOfBoundsException
+     *          If the preconditions on the {@code index}, {@code start},
+     *          and {@code end} parameters do not hold
+     *
+     * @throws  ReadOnlyBufferException
+     *          If this buffer is read-only
+     *
+     * @see #append(CharSequence,int,int)
+     * @see #put(int,$type$[],int,int)
+     * @see #put(int,$Type$Buffer,int,int)
+     * @since 21
+     */
+    public $Type$Buffer put(int index, CharSequence csq, int start, int end) {
+        Objects.checkFromToIndex(start, end, csq.length());
+        Objects.checkIndex(index, limit());
+        if (end - start > limit() - index)
+            throw new BufferOverflowException();
+        if (isReadOnly())
+            throw new ReadOnlyBufferException();
+
+        for (int i = start; i < end; i++)
+            this.put(index++, csq.charAt(i));
```

Comment:
This should probably include the `String::getBytes`, `StringBuffer::getBytes`, and `StringBuilder::getBytes` optimisation that `CharBuffer::append(CharSequece, int, int)` has as of [JDK‑8305811] ([GH‑13415]) and [JDK‑8306374] ([GH‑13522]).

[JDK‑8305811]: https://bugs.openjdk.org/browse/JDK-8305811 "[JDK‑8305811] (bf) Improve performance of CharBuffer::append(CharSequence[,int,int])"
[JDK‑8306374]: https://bugs.openjdk.org/browse/JDK-8306374 "[JDK‑8306374] (bf) Improve performance of DirectCharBuffer::append(CharSequence[,int,int])"
[GH‑13415]: https://github.com/openjdk/jdk/pull/13415
[GH‑13522]: https://github.com/openjdk/jdk/pull/13522

Reply:
> This should probably include the `String::getBytes`, `StringBuffer::getBytes`, and `StringBuilder::getBytes` optimisation that `CharBuffer::append(CharSequece, int, int)` has as of [JDK‑8305811](https://bugs.openjdk.org/browse/JDK-8305811) ([GH‑13415](https://github.com/openjdk/jdk/pull/13415)) and [JDK‑8306374](https://bugs.openjdk.org/browse/JDK-8306374) ([GH‑13522](https://github.com/openjdk/jdk/pull/13522)).

Clearly. That's why this is a draft PR. The optimizations are intentionally omitted until consensus is achieved on the specification.  I prefer not to redo the heap and direct overrides if it is at all avoidable.